From 8a1f4bf1b78e6e6853332073aaccb12d89e43f0d Mon Sep 17 00:00:00 2001
From: Wolfgang Mathurin <wmathurin@salesforce.com>
Date: Thu, 30 Nov 2017 14:00:21 -0800
Subject: [PATCH] Encrypt db if found unencrypted (5.3 bug fix)

---
 .../SmartStore/Classes/SFSmartStoreDatabaseManager.m   | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/libs/SmartStore/SmartStore/Classes/SFSmartStoreDatabaseManager.m b/libs/SmartStore/SmartStore/Classes/SFSmartStoreDatabaseManager.m
index 0fb06bae1..c511fdb6f 100644
--- a/libs/SmartStore/SmartStore/Classes/SFSmartStoreDatabaseManager.m
+++ b/libs/SmartStore/SmartStore/Classes/SFSmartStoreDatabaseManager.m
@@ -140,7 +140,25 @@ static NSString * const kSFSmartStoreVerifyReadDbErrorDesc = @"Could not read fr
     return [[self class] openDatabaseWithPath:fullDbFilePath key:key error:error];
 }
 
+// If you created your database with an app based on Mobile SDK 5.3.0 using cocoapod
+// Then the key was never applied, and the database can be read directly
+// This method checks for that situation and encrypt the database if needed
+- (void)fixFor530Bug:(NSString *)storeName key:(NSString *)key {
+    NSString *fullDbFilePath = [self fullDbFilePathForStoreName:storeName];
+    
+    __block BOOL needEncrypting = NO;
+    [[FMDatabaseQueue databaseQueueWithPath:fullDbFilePath] inDatabase:^(FMDatabase* db) {
+        needEncrypting = [[self class] verifyDatabaseAccess:db error:nil];
+    }];
+    
+    if (needEncrypting) {
+        [[self class] encryptDbWithStoreName:storeName storePath:fullDbFilePath key:key error:nil];
+    }
+}
+
 - (FMDatabaseQueue *)openStoreQueueWithName:(NSString *)storeName key:(NSString *)key error:(NSError **)error {
+    [self fixFor530Bug:storeName key:key];
+    
     __block BOOL result = YES;
     NSString *fullDbFilePath = [self fullDbFilePathForStoreName:storeName];
     FMDatabaseQueue *queue = [FMDatabaseQueue databaseQueueWithPath:fullDbFilePath];
-- 
2.13.2

